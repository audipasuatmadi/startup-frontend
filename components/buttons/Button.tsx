import React from 'react';
import BaseButton from './BaseButton';

type themes = 'brand' | 'add later';

export interface ButtonProps {
  disabled?: boolean;
  href?: string;
  children?: any;

  className?: string;
  type?: 'button' | 'submit' | 'reset';

  onClick?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;
  onMouseDown?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;

  rounded?: boolean;
  outlined?: boolean;
  theme?: themes;
}

/**
 * 
 * @version 0.1.0
 * @description basically wraps BaseButton, adding `outlined` and `rounded` features.
 * @param rounded boolean
 * @param outlined boolean
 * @author Putu Audi Pasuatmadi 
 */
const Button: React.FC<ButtonProps> = ({
  outlined,
  rounded,
  theme,
  disabled,
  className,
  ...otherProps
}: ButtonProps) => {
  const finalStyles = `button${rounded ? ' rounded-full' : ''}${
    disabled
      ? ` button-disabled bg-${theme}-dark text-white`
      : `${
          outlined
            ? ` bg-none border-2 border-${theme} text-${theme} hover:text-white`
            : ` bg-${theme} text-white`
        } hover:bg-${theme}-dark active:bg-${theme}-light focus:ring-4 ring-${theme} ring-opacity-50`
  } ${className}`;

  return (
    <BaseButton
      className={`${finalStyles}`}
      disabled={disabled}
      {...otherProps}
    />
  );
};

Button.defaultProps = {
  theme: 'brand'
}

export default Button;
