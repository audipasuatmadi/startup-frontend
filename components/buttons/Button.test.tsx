import React from 'react'
import { ShallowWrapper, shallow } from 'enzyme';
import Button from './Button';

const dummyText = 'test text';

describe('button rendering test', () => {
  let wrapper: ShallowWrapper;
  afterEach(() => {
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly without args & no theme', () => {
    wrapper = shallow(<Button />);
  });

  it('renders correctly as a Link', () => {
    wrapper = shallow(<Button href='/' />);
  });

  it('renders correctly when disabled', () => {
    wrapper = shallow(<Button disabled />);
  });

  it('renders correctly as a Link when disabled', () => {
    wrapper = shallow(<Button disabled href='/' />);
  });

  it('renders correctly with a children', () => {
    wrapper = shallow(<Button>{dummyText}</Button>);
  });

  it('renders correctly as a Link with a children', () => {
    wrapper = shallow(<Button href='/'>{dummyText}</Button>);
  });

  it('renders correctly with brand theme', () => {
    wrapper = shallow(<Button theme='brand' />);
  });

  it('renders correctly when rounded', () => {
    wrapper = shallow(<Button rounded />);
  });
});