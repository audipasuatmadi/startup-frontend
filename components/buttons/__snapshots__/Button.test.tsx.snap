// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button rendering test renders correctly as a Link 1`] = `
<BaseButton
  className="button bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
  href="/"
/>
`;

exports[`button rendering test renders correctly as a Link when disabled 1`] = `
<BaseButton
  className="button button-disabled bg-brand-dark text-white undefined"
  disabled={true}
  href="/"
/>
`;

exports[`button rendering test renders correctly as a Link with a children 1`] = `
<BaseButton
  className="button bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
  href="/"
>
  test text
</BaseButton>
`;

exports[`button rendering test renders correctly when disabled 1`] = `
<BaseButton
  className="button button-disabled bg-brand-dark text-white undefined"
  disabled={true}
/>
`;

exports[`button rendering test renders correctly when rounded 1`] = `
<BaseButton
  className="button rounded-full bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
/>
`;

exports[`button rendering test renders correctly with a children 1`] = `
<BaseButton
  className="button bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
>
  test text
</BaseButton>
`;

exports[`button rendering test renders correctly with brand theme 1`] = `
<BaseButton
  className="button bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
/>
`;

exports[`button rendering test renders correctly without args & no theme 1`] = `
<BaseButton
  className="button bg-brand text-white hover:bg-brand-dark active:bg-brand-light focus:ring-4 ring-brand ring-opacity-50 undefined"
/>
`;
